### Import tools

In [3]:
import gzip 
import tracemalloc


### Open file 

In [4]:
tracemalloc.start()

chr22_file_handle = gzip.open("/home/shatcher1/projects/PrimateAI-3D/data/PrimateAI-3D.hg38.chr22.txt.gz", "rt")

In [5]:
header = chr22_file_handle.readline().strip().split("\t")

In [6]:
print(header)

['chr', 'pos', 'non_flipped_ref', 'non_flipped_alt', 'gene_name', 'change_position_1based', 'ref_aa', 'alt_aa', 'score_PAI3D', 'percentile_PAI3D', 'refseq', 'prediction']


## Building Dictionary

In [7]:
variant_dict = {}
for line in chr22_file_handle:
    fields = line.strip().split("\t")                   # .strip() removes whitespace (spaces,tabs,newlines) from
    key = (fields[0], fields[1], fields[2], fields[3])  # from the beginning and end of a string.
    variant_dict[key] = fields[4:]                      # .split("\t") splits on tabs in the data, so one line in 
                                                        # "line" becomes a list

current, peak = tracemalloc.get_traced_memory()
tracemalloc.stop()                                                       

## Checking on Memory Usage

In [8]:
print(f"Peak memory usage: {peak / (1024**3):.2f} GB")  # This will be referenced in the markdown cell below

Peak memory usage: 1.00 GB


### Peak memory usage is 1 GB. 

### Going Back to our original scores

From running:
`zcat PrimateAI-3D.hg38.txt.gz | wc -l`

It returned 70,667,467 — meaning there are $\approx$ 70 million variants included in the original data.

In [7]:
chr22_variants = len(variant_dict) # This value is referenced in the markdown cell below
print(chr22_variants)

1476838


### Memory Analysis: Chr22 Dictionary

From running `zcat PrimateAI-3D.hg38.txt.gz | wc -l` in the terminal, the full dataset contains 70,667,467 lines (including 1 header line), meaning there are $\approx$ 70 million variants across the entire genome.

Loading chromosome 22 into a Python dictionary yields 1,476,838 variants, confirmed by `len(variant_dict)`. This represents approximately 2% of the full genome:

$$\frac{1{,}476{,}838}{70{,}666{,}467} \approx 2.09\%$$

The peak memory usage for the chr22 dictionary was measured at 1.00 GB using `tracemalloc`. Solving for how much memory we expect to need for the whole genome:

$$\frac{1.00 \text{ GB}}{0.0209} \approx 47.8 \text{ GB}$$

Needless to say, that's too much for my laptop, however 1 GB is manageable.